### PR TITLE
rpcserver: Add handleEstimateStakeDiff test.

### PR DESCRIPTION
**Pull request by JoeGruffins**
Because handleEstimateStakeDiff calls chain.EstimateNextStakeDifficulty
up to four times, change testRPCChain.estimateNextStakeDifficulty to a
function so that handlers can specify more complex return values. In
TestHandleExtimateStakeDiff, have the test RPC chain return functions
that return results from a queue, allowing for multiple calls with
varying results.

Use go-spew to print test result equality errors so that values of
pointers are also printed.

### CODE REVIEW

File: rpcserverhandlers_test.go [hunk 1]
```diff
@@ -1933,6 +1950,8 @@ func TestHandleEstimateFee(t *testing.T) {
 }
 
 func TestHandleEstimateSmartFee(t *testing.T) {
+	t.Parallel()
+
```
Comment (JoeGruffins) [inline rpcserverhandlers_test.go:1954]: Srry, I snuck this one past you last review.

File: rpcserverhandlers_test.go [hunk 2]
```diff
@@ -1975,6 +1994,122 @@ func TestHandleEstimateSmartFee(t *testing.T) {
...
+			queue := validQueueFunc()
+			queue[3].err = errors.New("")
+			chain.estimateNextStakeDifficultyQueue = queue
+			return chain
+		}(),
+		wantErr: true,
+		errCode: dcrjson.ErrRPCInternal.Code,
+	}})
+}
+
```
Comment (rstaudt2) [inline rpcserverhandlers_test.go:2112]: Test cases look good, hitting every case.  It's probably worth defining 
```
testChainParams.StakeDiffWindowSize = 144
```
in `defaultChainParams` so that if that value changes in `chaincfg.MainNetParams()` it won't potentially break these tests.

File: rpcserverhandlers_test.go (hunk 1 above)
Comment (rstaudt2) [reply rpcserverhandlers_test.go:1954]: Noticed that as well afterward, thanks.

File: rpcserverhandlers_test.go [hunk 3]
```diff
@@ -3450,9 +3585,30 @@ func testRPCServerHandler(t *testing.T, tests []rpcTest) {
...
+				if want.Min != got.Min ||
+					want.Max != got.Max ||
+					want.Expected != got.Expected ||
+					wantUser != gotUser {
+					t.Errorf("%s\nwant: %+v with user value %v\n got: %+v with user value %v\n",
+						test.name, want, wantUser, got, gotUser)
+				}
+			default:
+				if !reflect.DeepEqual(result, test.result) {
+					t.Errorf("%s\nwant: %+v\n got: %+v\n", test.name, test.result, result)
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply rpcserverhandlers_test.go:3610]: I'm not sure I follow, although printing structs after failing with go-spew is a great idea, the special case is because `reflect.DeepEqual()` will compare pointer values and not the underlying `User` value the pointer points to. Unless I am mistaken.

[the remark this answers is not in the document]
Comment (JoeGruffins) [reply rpcserverhandlers_test.go:3610]: Actually should work, in docs `Pointer values are deeply equal if they are equal using Go's == operator or if they point to deeply equal values.` So maybe I was doing something weird, let me try again...

[the remark this answers is not in the document]
Comment (JoeGruffins) [reply rpcserverhandlers_test.go:3610]: Actually, working now. I must have been doing something wierd. Apologies. But go-spew will be an asset.

File: rpcserverhandlers_test.go [hunk 4]
```diff
@@ -236,9 +244,15 @@ func (c *testRPCChain) CountVoteVersion(version uint32) (uint32, error) {
...
+// EstimateNextStakeDifficulty returns a mocked estimated next stake difficulty.
+// The return values are stored as a queue so that it can be called multiple
+// times with different results.
 func (c *testRPCChain) EstimateNextStakeDifficulty(newTickets int64, useMaxTickets bool) (int64, error) {
-	return c.estimateNextStakeDifficulty, nil
+	diff, err := c.estimateNextStakeDifficultyQueue[0].diff, c.estimateNextStakeDifficultyQueue[0].err
+	if len(c.estimateNextStakeDifficultyQueue) > 1 {
+		c.estimateNextStakeDifficultyQueue = c.estimateNextStakeDifficultyQueue[1:]
+	}
+	return diff, err
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply rpcserverhandlers_test.go:255]: Yes that works too, but not if the method takes no parameters. This one does of course. But then mid/user returns would just be something random I guess, like `case newTickets == 3`? Although, FIFO has problems too, like requiring quite some thought when creating, as the order is important. The test can easily be broken by changing up the handler. Leaving for now, but no problem changing.

File: rpcserverhandlers_test.go (hunk 2 above)
Comment (JoeGruffins) [reply rpcserverhandlers_test.go:2112]: Would be surprised if that changed, but good idea nonetheless.

File: rpcserverhandlers_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rstaudt2) [reply rpcserverhandlers_test.go:256]: Since this is just a function, the test case could implement pretty much anything (return based on params as above, just return something hardcoded, or use a queue or other data structure such as your example). Sounds good to leave for now.

File: rpcserverhandlers_test.go (hunk 2 above)
Comment (rstaudt2) [reply rpcserverhandlers_test.go:2113]: Agreed, this and some of the other defaultChainParams are not likely to change, but I think it makes sense for consistency as well.

File: rpcserverhandlers_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply rpcserverhandlers_test.go:256]: Oh, I see.

[the remark this answers is not in the document]
Comment (JoeGruffins) [reply rpcserverhandlers_test.go]: Ok! You're right this is perfect. Sorry your explanation flew over my head the first time. Thanks for not giving up on me!

**Comment by JoeGruffins** [comment]
@rstaudt2 
Your idea was really good. Please look one more time.

**Comment by JoeGruffins** [comment]
Will rebase.

**Comment by rstaudt2** [comment]
@JoeGruffins Nice, latest update looks good to me

**Comment by davecgh** [comment]
> Should I change the pr description?

Yes, please.  I like them to match reality since they get linked in release notes.
